internal/db/p2p/replicator.go

```
@@ -297,18 +301,36 @@ func (p *P2P) pushLogToReplicators(lg event.Update) {
 	reps, exists := p.replicators[lg.CollectionID]
 	p.repMu.Unlock()
 
+	for _, handler := range p.pushHandlers {
+		if err := handler.HandlePushToReplicators(context.Background(), lg); err != nil {
+			log.ErrorE("Push handler failed", err,
+				corelog.String("DocID", lg.DocID),
+				corelog.String("CollectionID", lg.CollectionID))
+		}
+	}
+
 	if exists {
 		for peerID := range reps {
 			go func() {
 				ctx, cancel := context.WithTimeout(p.ctx, networkRequestTimeout)
 				defer cancel()
-				if _, err := p.replicatorProtocol.PushToReplicator(ctx, lg, peerID); err != nil {
+				pushLogReq := protocol.PushLogRequest{
+					DocID:        lg.DocID,
+					CID:          lg.Cid.Bytes(),
+					CollectionID: lg.CollectionID,
+					Creator:      p.host.ID(),
+					Block:        lg.Block,
+				}
+				if _, err := p.replicatorProtocol.SendRequest(ctx, pushLogReq, peerID); err != nil {
 					log.ErrorE(
 						"Failed pushing log",
 						err,
 						corelog.String("DocID", lg.DocID),
 						corelog.Any("CID", lg.Cid),
 						corelog.Any("PeerID", peerID))
+					if !lg.IsRetry {
+						_ = p.handleReplicatorFailure(ctx, peerID, lg.DocID)
```

todo: Please document why discarding this error is desirable

---

I just moved this stuff from other place. Added log